core/src/storage/seed_encryption.py

```diff
@@ -0,0 +1,217 @@
+from typing import List, Tuple
+
+from storage import cache_codec, cache_common
+from trezor import utils
+
+FIELDS_TO_ENCRYPT = [cache_common.APP_COMMON_SEED]
+if not utils.BITCOIN_ONLY:
+    FIELDS_TO_ENCRYPT += [
+        cache_common.APP_CARDANO_ICARUS_SECRET,
+        cache_common.APP_CARDANO_ICARUS_TREZOR_SECRET,
+    ]
+
+FIELDS_TO_ENCRYPT_SESSIONLESS = [cache_common.APP_COMMON_SEED_WITHOUT_PASSPHRASE]
+
+
+def is_empty(session: cache_codec.DataCache, fields_to_encrypt: List[int]) -> bool:
+    """
+    Checks if the session has no data set for the fields to encrypt.
+    """
+    for field in fields_to_encrypt:
+        if session.get(field):
+            return False
+    return True
+
+
+def get_seed_encryption_key(session_id: bytes) -> bytes:
+    """
+    Returns the seed encryption key for a given session ID.
+    The key is derived from the device secret and the session ID.
+    """
+    from typing import TYPE_CHECKING
+
+    from storage.device import get_device_secret
+
+    from apps.common.seed import Slip21Node
+
+    if TYPE_CHECKING:
+        from apps.common.paths import Slip21Path
+
+    device_secret = get_device_secret()
+    label = b"Seed encryption key"
+    path: Slip21Path = [label, session_id]
+    key_node = Slip21Node(device_secret)
+    key_node.derive_path(path)
+    return key_node.key()
+
+
+def chain_seed_values(
+    session: cache_codec.DataCache, fileds_to_encrypt: List[int]
+) -> bytes:
+    """
+    Creates a plaintext representation of the session's seed
+    and other relevant data for encryption.
+
+    :param session: The session containing the data to encrypt.
+    :param fileds_to_encrypt: The list of fields to include in the plaintext.
+    :return: A bytes object containing the concatenated values of the fields.
+    """
+    plaintext = bytearray()
+    for field in fileds_to_encrypt:
+        value = session.get(field)
+        if value:
+            plaintext += value
+    return bytes(plaintext)
+
+
+def parse_value_chain_to_cache(
+    ciphertext: bytes, session: cache_codec.DataCache, fields_to_encrypt: List[int]
+) -> None:
```

Comment:
It seems to me that using `chain_seed_values()` and `parse_value_chain_to_cache()` is unnecessarily complicated. Both in terms code and memory allocations. Rather than building a large bytearray and then cutting it back up, I think you can simply call `encrypt()` and `decrypt()` on each value in succession and it works the same.